examples/benchmark/benchmark-matmult.cpp

```diff
@@ -31,15 +32,74 @@ static void ggml_graph_compute_helper(std::vector<uint8_t> & buf, ggml_cgraph *
     ggml_graph_compute(graph, &plan);
 }
 
+#define QK8_0 32
+
+typedef struct {
+    uint16_t d;       // delta
+    int8_t qs[QK8_0]; // quants
+} block_q8_0;
+
 static float tensor_sum_elements(const ggml_tensor * tensor) {
-    double sum = 0;
+    double sum                  = 0;
+    float  floatvalue           = 0;
+    unsigned short shortvalue   = 0;
+
     if (tensor->type == GGML_TYPE_F32) {
         for (int j = 0; j < tensor->ne[1]; j++) {
             for (int k = 0; k < tensor->ne[0]; k++) {
-                sum += ((float *) tensor->data)[j*tensor->ne[0] + k];
+                sum += ((float *) tensor->data)[j * tensor->ne[0] + k];
+            }
+        }
+    }
+
+    if (tensor->type == GGML_TYPE_I8) {
+        for (int j = 0; j < tensor->ne[1]; j++) {
+            for (int k = 0; k < tensor->ne[0]; k++) {
+                sum += ((int8_t *) tensor->data)[j * tensor->ne[0] + k];
+            }
+        }
+    }
+
+    if (tensor->type == GGML_TYPE_F16) {
+        for (int j = 0; j < tensor->ne[1]; j++) {
+            for (int k = 0; k < tensor->ne[0]; k++) {
+                shortvalue = ((unsigned short *) tensor->data)[j * tensor->ne[0] + k];
+                floatvalue = GGML_FP16_TO_FP32(shortvalue);
+                sum        += floatvalue;
+            }
+        }
+    }
+
+    if (tensor->type == GGML_TYPE_Q8_0) {
+        block_q8_0 * quant_datas = (block_q8_0 *)tensor->data;
+#if 1
+        ggml_type_traits_t qtype = ggml_internal_get_type_traits(tensor->type);
+        float * float32 = (float*)malloc((tensor->ne[0] * tensor->ne[1]) * sizeof(float));
+        if (NULL == float32) {
+            printf("malloc failed\n");
+            return 0.0;
+        }
+        qtype.to_float(quant_datas, float32, tensor->ne[0] * tensor->ne[1]);
+        for (int j = 0; j < tensor->ne[1]; j++) {
+            for (int k = 0; k < tensor->ne[0]; k++) {
+                sum += float32[j * tensor->ne[0] + k];
             }
         }
+        free(float32);
```

Comment:
You can get rid of `malloc` and all instances of `tensor->ne`:

```cpp
std::vector<float> f32out(ggml_nelements(tensor));
ggml_type_traits_t qtype = ggml_internal_get_type_traits(tensor->type);
qtype.to_float((void *) tensor->data, f32out.data(), f32out.size());
for (float e : f32out) sum += e;
```

It also works with other types of want, so maybe we can replace this whole function with this version.

Please note that I didn't test it, so up to you to test / adapt.

Reply:
excellent!